teamcity: failed tests on master: testrace/TestStoreSplitBeginTxnPushMetaIntentRace

The following tests appear to have failed:

[#22276](https://teamcity.cockroachdb.com/viewLog.html?buildId=22276):

```
--- FAIL: testrace/TestStoreSplitBeginTxnPushMetaIntentRace (0.004s)
Race detected!
I160915 16:52:22.619886 20183 storage/engine/rocksdb.go:356  opening in memory rocksdb instance
I160915 16:52:22.624140 20183 gossip/gossip.go:230  initial resolvers: []
W160915 16:52:22.625006 20183 gossip/gossip.go:1041  no resolvers found; use --join to specify a connected node
I160915 16:52:22.625254 20183 gossip/gossip.go:264  NodeID set to 1
I160915 16:52:22.625892 20183 gossip/gossip.go:278  NodeDescriptor set to node_id:1 address:<network_field:"" address_field:"" > attrs:<> 
I160915 16:52:22.638672 20183 gossip/gossip.go:264  NodeID set to 1
I160915 16:52:22.646113 20234 storage/replica.go:3063  [s1,r1{/Min-/Max}] gossiping cluster id {"00000000-0000-0000-0000-000000000000"} from store 1, range 1
I160915 16:52:22.647462 20183 storage/store.go:1016  [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I160915 16:52:22.649586 20190 storage/replica.go:2455  [s1,r1{/Min-/Max}] applying command with forced error: storage/replica.go:2380: no-op on empty Raft entry
E160915 16:52:22.650297 20190 storage/replica.go:2490  [s1,r1{/Min-/Max}] error executing raft command: storage/replica.go:2380: no-op on empty Raft entry
I160915 16:52:22.651088 20190 storage/replica_trigger.go:249  [s1,r1{/Min-/Max}] new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 900.000123ms following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0s [physicalTime=1970-01-01 01:00:00.000000124 +0000 UTC]
I160915 16:52:22.658286 20234 storage/replica.go:3085  [s1,r1{/Min-/Max}] gossiping sentinel from store 1, range 1
I160915 16:52:22.659019 20234 storage/replica.go:3092  [s1,r1{/Min-/Max}] gossiping first range from store 1, range 1: [{1 1 1}]
I160915 16:52:22.659477 20215 kv/dist_sender.go:222  gossipped first range descriptor: [{NodeID:1 StoreID:1 ReplicaID:1}]
I160915 16:52:22.664686 20183 storage/replica_command.go:2291  [s1,r1{/Min-/Max}] initiating a split of this range at key /System/""
==================
WARNING: DATA RACE
Write at 0x00c4212ac860 by goroutine 120:
  github.com/cockroachdb/cockroach/storage_test.TestStoreSplitBeginTxnPushMetaIntentRace()
      /go/src/github.com/cockroachdb/cockroach/storage/client_split_test.go:1613 +0x5b1
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:610 +0xc9

Previous read at 0x00c4212ac860 by goroutine 115:
  github.com/cockroachdb/cockroach/storage.(*Replica).withRaftGroupLocked()
      /go/src/github.com/cockroachdb/cockroach/storage/replica.go:407 +0x2a2
  github.com/cockroachdb/cockroach/storage.(*Replica).handleRaftReady()
      /go/src/github.com/cockroachdb/cockroach/storage/replica.go:1767 +0x2fe
  github.com/cockroachdb/cockroach/storage.(*Store).processReady()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:2629 +0x14c
  github.com/cockroachdb/cockroach/storage.(*raftScheduler).worker()
      /go/src/github.com/cockroachdb/cockroach/storage/scheduler.go:197 +0x427
  github.com/cockroachdb/cockroach/storage.(*raftScheduler).Start.func1()
      /go/src/github.com/cockroachdb/cockroach/storage/scheduler.go:160 +0x4a
  github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1()
      /go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:187 +0x8a

Goroutine 120 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:646 +0x52f
  testing.RunTests.func1()
      /usr/local/go/src/testing/testing.go:793 +0xb9
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:610 +0xc9
  testing.RunTests()
      /usr/local/go/src/testing/testing.go:799 +0x4ba
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:743 +0x12f
  github.com/cockroachdb/cockroach/storage_test.TestMain()
      /go/src/github.com/cockroachdb/cockroach/storage/main_test.go:57 +0x286
  main.main()
      github.com/cockroachdb/cockroach/storage/_test/_testmain.go:720 +0x1b5

Goroutine 115 (running) created at:
  github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker()
      /go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x73
  github.com/cockroachdb/cockroach/storage.(*raftScheduler).Start()
      /go/src/github.com/cockroachdb/cockroach/storage/scheduler.go:161 +0x108
  github.com/cockroachdb/cockroach/storage.(*Store).processRaft()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:2677 +0xa8
  github.com/cockroachdb/cockroach/storage.(*Store).Start()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:905 +0x737
  github.com/cockroachdb/cockroach/storage_test.createTestStoreWithEngine()
      /go/src/github.com/cockroachdb/cockroach/storage/client_test.go:163 +0xcf2
  github.com/cockroachdb/cockroach/storage_test.createTestStoreWithContext()
      /go/src/github.com/cockroachdb/cockroach/storage/client_test.go:99 +0x263
  github.com/cockroachdb/cockroach/storage_test.TestStoreSplitBeginTxnPushMetaIntentRace()
      /go/src/github.com/cockroachdb/cockroach/storage/client_split_test.go:1599 +0x16c
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:610 +0xc9
==================
I160915 16:52:22.699984 20175 storage/client_split_test.go:1637  splitting at "a"
E160915 16:52:22.702038 20201 storage/queue.go:542  [replicate] (purgatory) on [n1,s1,r1{/Min-/System/""}]: 0 of 0 stores with an attribute matching []; likely not enough nodes in cluster
I160915 16:52:22.703143 20227 storage/replica.go:2455  [s1,r2{/System/""-/Max}] applying command with forced error: storage/replica.go:2380: no-op on empty Raft entry
I160915 16:52:22.703383 20175 storage/replica_command.go:2291  [s1,r2{/System/""-/Max}] initiating a split of this range at key "a"
E160915 16:52:22.704050 20227 storage/replica.go:2490  [s1,r2{/System/""-/Max}] error executing raft command: storage/replica.go:2380: no-op on empty Raft entry
E160915 16:52:22.705743 20201 storage/queue.go:542  [replicate] (purgatory) on [n1,s1,r2{/System/""-/Max}]: 0 of 0 stores with an attribute matching []; likely not enough nodes in cluster
I160915 16:52:22.706673 20228 storage/replica_command.go:85  [s1,r2{/System/""-/Max}] test injecting error: node unavailable; try another peer
I160915 16:52:22.707382 20175 kv/dist_sender.go:1028  application error: node unavailable; try another peer
I160915 16:52:22.707708 20175 kv/dist_sender.go:759  reply error [txn: 96fde23d], BeginTransaction [/Local/Range/"\x04"/RangeDescriptor,/Min), ConditionalPut [/Local/Range/"\x04"/RangeDescriptor,/Min): failed to send RPC: sending to all 1 replicas failed; last error: node unavailable; try another peer
I160915 16:52:22.760083 20191 storage/client_split_test.go:1621  allowing BeginTransaction to proceed after 20ms
I160915 16:52:22.770202 20183 storage/gc_queue.go:317  [gc] completed with stats {Now:3600.000000124,97 Policy:{TTLSeconds:86400} GCKeys:0 IntentsConsidered:0 IntentTxns:0 TransactionSpanTotal:1 TransactionSpanGCAborted:0 TransactionSpanGCCommitted:0 TransactionSpanGCPending:0 AbortSpanTotal:0 AbortSpanConsidered:0 AbortSpanGCNum:0 PushTxn:0 ResolveTotal:0 ResolveSuccess:0 Threshold:-82799.-999999876,0}
I160915 16:52:22.779474 20226 storage/replica.go:2455  [s1,r3{"a"-/Max}] applying command with forced error: storage/replica.go:2380: no-op on empty Raft entry
E160915 16:52:22.780365 20226 storage/replica.go:2490  [s1,r3{"a"-/Max}] error executing raft command: storage/replica.go:2380: no-op on empty Raft entry
E160915 16:52:22.781022 20201 storage/queue.go:542  [replicate] (purgatory) on [n1,s1,r3{"a"-/Max}]: 0 of 0 stores with an attribute matching []; likely not enough nodes in cluster
I160915 16:52:22.782409 20183 storage/gc_queue.go:317  [gc] completed with stats {Now:3600.000000124,112 Policy:{TTLSeconds:86400} GCKeys:0 IntentsConsidered:0 IntentTxns:0 TransactionSpanTotal:0 TransactionSpanGCAborted:0 TransactionSpanGCCommitted:0 TransactionSpanGCPending:0 AbortSpanTotal:0 AbortSpanConsidered:0 AbortSpanGCNum:0 PushTxn:0 ResolveTotal:0 ResolveSuccess:0 Threshold:-82799.-999999876,0}
I160915 16:52:22.785111 20183 stopper.go:408  quiesceing; tasks left:
1      storage/intent_resolver.go:371
I160915 16:52:22.790918 20183 storage/engine/rocksdb.go:404  closing in-memory rocksdb instance

```
Please assign, take a look and update the issue accordingly.

--- a/storage/client_split_test.go
+++ b/storage/client_split_test.go
@@ -52,6 +52,7 @@ import (
 	"github.com/cockroachdb/cockroach/util/log"
 	"github.com/cockroachdb/cockroach/util/protoutil"
 	"github.com/cockroachdb/cockroach/util/randutil"
+	"github.com/cockroachdb/cockroach/util/syncutil"
 	"github.com/cockroachdb/cockroach/util/timeutil"
 	"github.com/cockroachdb/cockroach/util/tracing"
 	"github.com/coreos/etcd/raft/raftpb"
@@ -1594,27 +1595,28 @@ func TestStoreSplitBeginTxnPushMetaIntentRace(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	splitKey := roachpb.RKey("a")
 
-	sCtx := storage.TestStoreContext()
-	sCtx.TestingKnobs.DisableSplitQueue = true
-	store, stopper, manual := createTestStoreWithContext(t, sCtx)
-	defer stopper.Stop()
-
-	// Advance the clock past the transaction cleanup expiration.
-	manual.Increment(storage.GetGCQueueTxnCleanupThreshold().Nanoseconds() + 1)
-
-	// First, create a split after addressing records.
-	args := adminSplitArgs(roachpb.KeyMin, keys.SystemPrefix)
-	if _, pErr := client.SendWrapped(rg1(store), nil, &args); pErr != nil {
-		t.Fatal(pErr)
+	var startMu struct {
+		syncutil.Mutex
+		time time.Time
 	}
 
-	startTime := timeutil.Now()
 	wroteMeta2 := make(chan struct{}, 1)
-	store.TestingKnobs().TestingCommandFilter = func(filterArgs storagebase.FilterArgs) *roachpb.Error {
+
+	sCtx := storage.TestStoreContext()
+	sCtx.TestingKnobs.DisableSplitQueue = true
+	sCtx.TestingKnobs.TestingCommandFilter = func(filterArgs storagebase.FilterArgs) *roachpb.Error {
+		startMu.Lock()
+		start := startMu.time
+		startMu.Unlock()
+
+		if start.IsZero() {
+			return nil
+		}
+
 		if _, ok := filterArgs.Req.(*roachpb.BeginTransactionRequest); ok {
 			// Return a node unavailable error for BeginTransaction request
 			// in the event we haven't waited 20ms.
-			if timeutil.Since(startTime) < 20*time.Millisecond {
+			if timeutil.Since(start) < 20*time.Millisecond {
 				return roachpb.NewError(&roachpb.NodeUnavailableError{})
 			}
 			if log.V(1) {
@@ -1629,6 +1631,22 @@ func TestStoreSplitBeginTxnPushMetaIntentRace(t *testing.T) {
 		}
 		return nil
 	}
+	store, stopper, manual := createTestStoreWithContext(t, sCtx)
+	defer stopper.Stop()
+
+	// Advance the clock past the transaction cleanup expiration.
+	manual.Increment(storage.GetGCQueueTxnCleanupThreshold().Nanoseconds() + 1)
+
+	// First, create a split after addressing records.
+	args := adminSplitArgs(roachpb.KeyMin, keys.SystemPrefix)
+	if _, pErr := client.SendWrapped(rg1(store), nil, &args); pErr != nil {
+		t.Fatal(pErr)
+	}
+
+	// Begin blocking BeginTransaction calls and signaling meta2 writes.
+	startMu.Lock()
+	startMu.time = timeutil.Now()
+	startMu.Unlock()
 
 	// Initiate split at splitKey in a goroutine.
 	doneSplit := make(chan *roachpb.Error, 1)

